### PR TITLE
[ENG-5551] Phase 4: Upgrade to python v3.12

### PR DESCRIPTION
**Pull request by opaduchak**
## Purpose

Upgrade python to 3.12

## Changes

Upgraded python in Dockerfile, docker-compose, github actions, fixed all related issues:
- fixed and updated pre-commit hooks
- removed asynctest
- replaced boto with boto3
- replaced usages of distutils with packaging
- upgraded CI for python 3.12
- upgraded docker-compose.yml to python 3.12
- changed base image to python:3.12-alpine3.17
- removed recent crutches to get old versions of dependencies to install on python 3.10
- fixed other nuances

## QA Notes

**verify s3 addon works as expected**

## Ticket
https://openscience.atlassian.net/browse/ENG-5551

### CODE REVIEW

File: .github/actions/start-build/action.yml
Comment (cslzchen) [inline .github/actions/start-build/action.yml:40]: Should we still ceiling or ping setup-tools with a higher version?

File: Dockerfile
Comment (cslzchen) [inline Dockerfile:1]: For @felliott: now we use the python 3.12 image which installs node 18 by default

File: addons/s3/requirements.txt
Comment (cslzchen) [inline addons/s3/requirements.txt:1]: For @felliott : this is something we were concerned about but everything seems fine with Oleh's fix, at least on the OSF side; we still need to locally or staging2 test this with WB S3 (which still uses `boto`).

File: addons/s3/views.py
Comment (cslzchen) [inline addons/s3/views.py:148]: I am wondering if `str(e)` would reveal sensitive data? It depends on how `NoCredentialsError` implements it. Please double check.

File: admin_tests/users/test_views.py
Comment (cslzchen) [inline admin_tests/users/test_views.py:589]: For this test, I didn't see why the original tests was wrong previous. We probably need to have a list of tests with same/similar fixes and go through each one by one.

File: docker-compose.yml
Comment (cslzchen) [inline docker-compose.yml:36]: Since we did this, I think we can remove all
```yaml
# - osf_requirements_vol:/usr/local/lib/python3.6
```
in this file. A total of 7 occurrences if my count is correct.

File: osf_tests/test_archiver.py
Comment (cslzchen) [inline osf_tests/test_archiver.py:482]: I think I might be wrong during our discussion, I believe the `mock_file_tree` and `mock_group` are accessible outside the `with` context. See this [link](https://stackoverflow.com/questions/6432355/variable-defined-with-with-statement-available-outside-of-with-block/6432414#6432414). So, it seems to be a correct test before the upgrade.

Please take a look at all similar/same cases that are skipped.

File: requirements.txt
Comment (cslzchen) [inline requirements.txt:39]: I am not familiar with `pymongo`, any details and/or references on this change?

**Review by cslzchen — CHANGES_REQUESTED** [review]
First pass done and great work 👍 👍 👍 .

In addition to the comments I left:

- [x] I love the new style fixes, is this automatically done since we now have a working pre-commit? If so, ignore my next sentence. If not, I would suggest doing style change in  a separate PR so it's easier for CR to focus on the functionality code upgrade. I would expect not only upgrade-related merge conflicts but also massive style change related ones  we merge new `develop`.

- [x] In addition, I suggest we don't skip the failed tests due to `assert_called_with()` change in this PR. Let them fail and we can create another ticket to look into it deeper.

- [x] Finally, @felliott and/or I need to verify `boto3` with WB locally. Maybe we use this weeks ForestFire to setup WB and connect S3 locally together?

File: Dockerfile
Comment (felliott) [reply Dockerfile:1]: huzzah!

File: requirements.txt
Comment (opaduchak) [reply requirements.txt:39]: This change probably has been lost in some merge, but it fixes the problem of incompatibility between mongodb 3.4 and python 3.10+

File: osf_tests/test_archiver.py
Comment (opaduchak) [reply osf_tests/test_archiver.py:481]: The root of the issue is not related to context manager. The only thing which changes after an exit from context manager is that patch is undone. Mock object retains all the information about calls executed on it after closure of patch's context manager.

File: admin_tests/users/test_views.py
Comment (opaduchak) [reply admin_tests/users/test_views.py:588]: Previous tests were wrong because mocks do not have a method called `called_with`.
I'll provide image with illustrates the issue the best

image

File: .github/actions/start-build/action.yml
Comment (opaduchak) [reply .github/actions/start-build/action.yml:40]: I wouldn't say so, the only reason for `setuptools` ceiling was that old `python3.6` dependencies refused to install on `python 3.10` without `setuptools` constrained `<58.0.0`

File: addons/s3/views.py
Comment (opaduchak) [reply addons/s3/views.py:148]: It won't, double-checked

**Comment by opaduchak** [comment]
> I love the new style fixes, is this automatically done since we now have a working pre-commit? If so, ignore my next sentence. If not, I would suggest doing style change in a separate PR so it's easier for CR to focus on the functionality code upgrade. I would expect not only upgrade-related merge conflicts but also massive style change related ones we merge new develop.

New style fixes are from pre-commit hooks, I've also run pyupgrade for python 3.12, but it chenged only less than 10 files

File: admin_tests/users/test_views.py
Comment (cslzchen) [reply admin_tests/users/test_views.py:588]: ah, I see thanks for the explanation, I didn't get this point during our meeting

File: osf_tests/test_archiver.py
Comment (cslzchen) [reply osf_tests/test_archiver.py:481]: ah, make sense

**Review by cslzchen — APPROVED** [review]

File: Dockerfile
Comment (felliott) [inline Dockerfile:82]: Not a dealbreaker for me, but I think the `|| true` was there to make it so the image can still build for local dev even if one of the requirements fails to install properly. An affordance to help troubleshoot / fix those exact sort of issues.  But my brain is getting fuzzy, I'll leave it up to y'all.

**Review by felliott — CHANGES_REQUESTED** [review]
Hey @opaduchak!

This is looking pretty solid!  I have a few comments, a few nitpicks, and a few questions. But since I won't be back until Monday, @cslzchen feel free to go ahead and merge even if not everything has been addressed. We can always fix stuff in followup tickets.

Cheers,
@felliott

File: addons/s3/utils.py
Comment (opaduchak) [reply addons/s3/utils.py:123]: boto3 uses `OrdinaryCallingFormat` by default, so this should not be an issue

File: Dockerfile
Comment (opaduchak) [reply Dockerfile:82]: Well, from what I see, we do not use this Dockerfile for development purposes, and it may  backfire us if this occurs in release build (compilation failed but we move on). Moreover, as I understand we have `requirements`  service in `docker-compose.yml` which is responsible for building dependencies for dev environment. Please, correct me if I'm mistaken

File: docker-compose.yml
Comment (opaduchak) [reply docker-compose.yml:369]: Okay, this makes makes sense, I forgot that we use `werkzeug` instead of `gunicorn` for dev env

File: docker-compose.yml
Comment (opaduchak) [reply docker-compose.yml:382]: This change is related to change of python lib location in `python:3.12-alpine3.17` related to `node:18-alpine3.17` docker images

File: api/base/utils.py
Comment (opaduchak) [reply api/base/utils.py:3]: I also thought so, but this change kinda had a history:
When I found that disutils are missing in 3.12 I found a replacement (`packaging`) and was happy that it was already installed, but it seems it is installed only along with some dev deps.
Consequently, maybe we'd be better off using `semver` package which is desingned to do only this thing

File: api/preprints/serializers.py
Comment (cslzchen) [reply api/preprints/serializers.py]: This is update is automatically done by pre-commit I believe.

File: api/preprints/views.py
Comment (cslzchen) [reply api/preprints/views.py]: See above

File: Dockerfile
Comment (cslzchen) [reply Dockerfile:82]: * I agree with Oleh. We use the already build image locally
* I think we can remove comments in line 81 too, see
  * https://github.com/CenterForOpenScience/osf.io/commit/fc36671a52228b7afdfdb58c64ecaa492edab032
  * https://github.com/CenterForOpenScience/osf.io/pull/6783

File: api/base/utils.py
Comment (cslzchen) [reply api/base/utils.py]: @felliott, our versioning is not valid SemVer style, switching to `semver` breaks the API

> ValueError: 2.0 is not valid SemVer string

In addition, I think `packaging` is fine, we should expect user to provide the correct version number. Although our comparison not only includes `>`, `<` but also `>=` and `<=`, it's fine with `packaging`.

```
from packaging.version import Version
>>> Version('1.2.8.5') <= Version('1.2')
False
>>> Version('1.2.8.5') <= Version('1.2.8.6')
True
```

**Review by cslzchen — APPROVED** [review]
Looks good, leaving three comments unresolved for @felliott 's awareness of our discussion/decision. Will merge after CI checked. 📦

**Pull request merged**
